File: homework4_1/src/main/java/com/example/homework4_1/ContactsAdapter.java
package com.example.homework4_1;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.homework4_1.Contact.ConnectType;
import com.example.homework4_1.Contact.Contact;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

public class ContactsAdapter extends RecyclerView.Adapter<ContactsAdapter.ItemViewHolder> implements Filterable {

    private List<Contact> contactsLocal = new ArrayList<>();
    private List<Contact> contactsAll = new ArrayList<>();
    private MainActivity.ListItemActionListener listItemActionListener;


    public ContactsAdapter(List<Contact> contacts, MainActivity.ListItemActionListener listItemActionListener) {
        contactsLocal.addAll(contacts);
        contactsAll.addAll(contacts);
        this.listItemActionListener = listItemActionListener;
    }

    public void setContacts(List<Contact> contacts) {
        contactsLocal.clear();
        contactsLocal.addAll(contacts);
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public ItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.contact, parent, false);
        return new ItemViewHolder(view, listItemActionListener);
    }

    @Override
    public void onBindViewHolder(@NonNull ItemViewHolder holder, int position) {
        holder.bind(contactsLocal.get(position));

    }

    @Override
    public int getItemCount() {
        return contactsLocal != null ? contactsLocal.size() : 0;
    }

    @Override
    public Filter getFilter() {
        return contactsFilter;
    }

    private Filter contactsFilter = new Filter() {
        @Override
        protected FilterResults performFiltering(CharSequence constraint) {
            List<Contact> filteredContacts = new ArrayList<>();
            if (constraint == null || constraint.length() == 0) {
                filteredContacts.addAll(contactsAll);
            } else {
                String filterPattern = constraint.toString().toLowerCase().trim();
                for (Contact contact : contactsAll) {
                    if (contact.getName().toLowerCase().contains(filterPattern)) {
                        filteredContacts.add(contact);

                    }
                }
            }
            FilterResults results = new FilterResults();
            results.values = filteredContacts;
            return results;
        }

        @Override
        protected void publishResults(CharSequence constraint, FilterResults results) {
            contactsLocal.clear();
            contactsLocal.addAll((Collection<? extends Contact>) results.values);
            notifyDataSetChanged();
        }
    };

    static class ItemViewHolder extends RecyclerView.ViewHolder {

        LinearLayout contactElement;
        ImageView contactIcon;
        TextView contactName;
        TextView contactCommunication;
        private MainActivity.ListItemActionListener listItemActionListener;

        public ItemViewHolder(@NonNull View itemView, MainActivity.ListItemActionListener listItemActionListener) {
            super(itemView);

            contactElement = itemView.findViewById(R.id.contact_element);
            contactIcon = itemView.findViewById(R.id.contact_icon);
            contactName = itemView.findViewById(R.id.contact_name);
            contactCommunication = itemView.findViewById(R.id.contact_communication);
            this.listItemActionListener = listItemActionListener;
        }

        public void bind(final Contact contact) {

            contactElement.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (listItemActionListener != null) {
                        listItemActionListener.onItemClicked(contact);
                    }
                }
            });
            if (contact.getConnectType() == ConnectType.EMAIL) {
                this.contactIcon.setImageResource(R.drawable.ic_baseline_contact_mail_24);
            } else {
                this.contactIcon.setImageResource(R.drawable.ic_baseline_contact_phone_24);
            }
            this.contactName.setText(contact.getName());
            this.contactCommunication.setText(contact.getCommunication());

        }
    }

}

File: homework5_2/src/main/java/com/example/homework5_2/ContactsAdapter.kt
package com.example.homework5_2

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filterable
import android.widget.Filter
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView

import androidx.recyclerview.widget.RecyclerView
import com.example.homework5_2.Contact.ConnectType
import com.example.homework5_2.Contact.Contact
import java.util.*

interface ListItemActionListener {
    fun onItemClicked(contact: Contact?)
}

class ContactsAdapter(
    contacts: List<Contact>?,
    listItemActionListener: ListItemActionListener
) :
    RecyclerView.Adapter<ContactsAdapter.ItemViewHolder>(), Filterable {
    private val contactsLocal = mutableListOf<Contact>()
    private val contactsAll = mutableListOf<Contact>()
    private var listItemActionListener: ListItemActionListener
        
    init {
        if(contacts != null) {
            contactsLocal.addAll(contacts)
            contactsAll.addAll(contacts)
        }
        this.listItemActionListener = listItemActionListener
    }
    public fun setContacts(contacts: List<Contact>?) {
        contactsAll.clear()
        contactsLocal.clear()
        if (contacts != null) {
            contactsAll.addAll(contacts)
            contactsLocal.addAll(contacts)
        }

        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.contact, parent, false)
        return ItemViewHolder(view, listItemActionListener)
    }

    override fun onBindViewHolder(holder: ItemViewHolder, position: Int) {
        contactsLocal[position].let { holder.bind(it) }
    }

    override fun getItemCount(): Int {
        return contactsLocal.size ?: 0
    }

    override fun getFilter(): Filter {
        return contactsFilter
    }

    private val contactsFilter: Filter = object : Filter() {
        override fun performFiltering(constraint: CharSequence): FilterResults {
            val filteredContacts = mutableListOf<Contact>()
            if (constraint.isEmpty()) {
                filteredContacts.addAll(contactsAll)
            } else {
                val filterPattern = constraint.toString().toLowerCase().trim { it <= ' ' }
                for (contact in contactsAll) {
                    if (contact.name.toLowerCase().contains(filterPattern)) {
                        filteredContacts.add(contact)
                    }
                }
            }
            val results = FilterResults()
            results.values = filteredContacts
            return results
        }

        override fun publishResults(constraint: CharSequence, results: FilterResults) {
            contactsLocal.clear()
            contactsLocal.addAll(results.values as MutableList<Contact>)
            notifyDataSetChanged()
        }
    }

    class ItemViewHolder(
        itemView: View,
        listItemActionListener: ListItemActionListener?
    ) :
        RecyclerView.ViewHolder(itemView) {
        var contactElement: LinearLayout = itemView.findViewById(R.id.contact_element)
        var contactIcon: ImageView = itemView.findViewById(R.id.contact_icon)
        var contactName: TextView = itemView.findViewById(R.id.contact_name)
        var contactCommunication: TextView = itemView.findViewById(R.id.contact_communication)
        private val listItemActionListener: ListItemActionListener? =
            listItemActionListener

        fun bind(contact: Contact) {
            contactElement.setOnClickListener { listItemActionListener?.onItemClicked(contact) }
            val resId = if (contact.connectType == ConnectType.EMAIL) {
                R.drawable.ic_baseline_contact_mail_24
            } else {
                R.drawable.ic_baseline_contact_phone_24
            }
            contactIcon.setImageResource(resId)
            contactName.text = contact.name
            contactCommunication.text = contact.communication
        }

    }

    fun addItem(contact: Contact) {
        contactsAll.add(contact)
        contactsLocal.add(contact)
        notifyDataSetChanged()
    }
}


